### ISSUE TITLE
 Implement Dask equivalent of pandas.io.json.json_normalize

### ISSUE THREAD

**Issue by nsadeh** [dask/dask#4170]
pandas has a useful function to flatten json files into a DataFrame structure : https://pandas.pydata.org/pandas-docs/stable/generated/pandas.io.json.json_normalize.html

Would be nice to have this feature in Dask as well. I use Dask but am not a heavy power-user, I tried to implement this using map_partitions but got stuck in creating an appropriate meta parameter. 

It should hopefully be easy to recreate this: https://www.kaggle.com/jboysen/quick-tutorial-flatten-nested-json-in-pandas# in Dask.

**Comment by martindurant** [comment]
Have you tried looking at dask.map's `map` and `pluck`?

**Comment by nsadeh** [comment]
Looking now. Have never used it before.

I have a DataFrame where some columns are json and I am trying to flatten them. Some of the json columns, when flattened, could have their own json columns.

Basically, would I be casting the column (Series) into a bag, then mapping (using bag.map) and use pluck to define the column names, then to_dataframe? And applying this all to the dataframe partitions?

**Comment by mlisovyi** [comment]
Is there a working example of how to implement it? I'm 100% a beginner with dask, but bumped into the same situation

**Comment by ap1495** [comment]
I'm new to Dask as well and would love a function similar to pandas json_normalize..
For now could you please share a working example of how we can achieve this using dask.bag?

Thanks !!

**Comment by mlisovyi** [comment]
I managed to implement it and it is available in [this kernel on kaggle](https://www.kaggle.com/mlisovyi/bigdata-dask-pandas-flat-json-trim-data-upd?scriptVersionId=7750362). 

The essence is that you do something like:
`df[new_column_name] = df[json_field].to_bag().pluck(key, default=default_value).to_dataframe().iloc[:,0]`,
where `json_field` is the column in a dask dataframe containing dictionaries (you have to transform strings (which you will most likely have, if you read in data from a file) into python dictionaries), and you iterate over `key` in the dictionary providing the `default_value` in case of you have incomplete population of dictionaries between rows

**Comment by jakirkham** [comment]
Glad to hear you got something working, @mlisovyi. Given the interest here, I wonder if there is a general function that would be useful. Maybe worth a PR? Any thoughts on this @TomAugspurger?

**Comment by martindurant** [comment]
I think we can safely close this, as there seems to be a workaround, but no general framework we can propose for the variety of JSON input that users may be interested in.

**Comment by TomAugspurger** [comment]
Agreed with @martindurant that this would be difficult to solve in general. For simple cases, `dask.delayed` works well.

```python
>>> data = [{'id': 1, 'name': {'first': 'Coleen', 'last': 'Volk'}},
...         {'name': {'given': 'Mose', 'family': 'Regner'}},
...         {'id': 2, 'name': 'Faye Raker'}]
>>> big_data = [data, data, data]
>>> dfs = [dask.delayed(pd.io.json.json_normalize)(x) for x in big_data]
>>> df = dd.from_delayed(dfs)
```

For anything more complex, moving to the dask.bag API is reasonable. Documentation writing around this would be valuable.

**Comment by manojonline** [comment]
> dfs = [dask.delayed(pd.io.json.json_normalize)(x) for x in big_data]

The json_normalize as above seems to put the entire work load on a single worker. When loading big datasets, there is a risk of worker running out of memory?

**Comment by TomAugspurger** [comment]
> The json_normalize as above seems to put the entire work load on a single worker

Why do you think that? It should result in one task per Delayed result in `dfs`. If the `x` passed to `delayed(json_normalize)` is too large then you would indeed have memory issues.

**Comment by manojonline** [comment]
I tried something similar, and i saw that the entire json_normalize was
passed onto one worker.
This was my setup.

import json
import dask
import pandas as pd
import dask.bag as db
import dask.dataframe as dd
from pandas.io.json import json_normalize

bag = db.read_text('*.json').map(json.loads)

json_list = [dask.delayed(pd.io.json.json_normalize)(bag,
record_path=['shops', ['items']],
                                                 meta=[['shops',
'creation_date'], ['shops', 'owner'],
                                                       ['shops', 'title'],
['shops', 'type'],
                                                       ['shops',
'location', 'map'], ['shops', 'location', 'x'],
                                                       ['shops',
'location', 'y'], 'generation_timestamp'])]

df = dd.from_delayed(json_list)


This Dask Graph showed that the entire json_normalize was sent to a single
worker and it ran into memory issues.
The directory has over 11000 json files.

PS: I'm very new to Python as well dask and pandas.

Cheers
Manoj

**Comment by kevinschaich** [comment]
Hey guys – may have found a solution that works w/ the default distributed readers + map functions:

```python
df = df.map_partitions(lambda x: pd.json_normalize(x.to_dict(orient="records")), enforce_metadata=False)
```

Before & after:

<img width="1289" alt="image" src="https://user-images.githubusercontent.com/9244728/218260004-c110abbb-33c4-4518-91da-c5ffa491ba66.png">

Full example:

```python
import dask.dataframe as dd
import pandas as pd

# assuming your JSON is standard (array of objects, comma-separated, not ND-JSON)
df = dd.read_json("/path/to/my/data.json", orient="records", lines=False)

df = df.map_partitions(lambda x: pd.json_normalize(x.to_dict(orient="records")), enforce_metadata=False)
```

Let me know if this is the most efficient way to do it / if anyone has suggested improvements. I tried getting Bags to work with `.map` but didn't have any luck.

**Comment by kevinschaich** [comment]
Upon some further testing and attempting to write this DataFrame back to disk (I've tried json, csv, parquet), I'm encountering some schema errors: 

```
    tracks.href: string
    tracks.total: int64

This error *may* be resolved by passing in schema information for
the mismatched column(s) using the `schema` keyword in `to_parquet`.
```

Let me know if anyone has suggestions.

**Comment by martindurant** [comment]
Your errors will happen if there are sections of your data which have no valid values. You are best including a `meta=` to `map_partitions`, with a value being a zero-length dataframe of the correct dtypes expected for your data (with href maybe as object and total maybe as Int64 or float64). I think this is probably what you intended `enforce_metadata` for, but that acts for the case that you provide a meta.

Otherwise, the technique is fine, and map_partitions often does play this role of plugging in for functions not already provided by dask.

**Comment by kevinschaich** [comment]
@martindurant I tried a version of that:

```python
schema = pd.json_normalize(df.partitions[0].head(1).to_dict(orient="records"))
df = df.map_partitions(lambda x: pd.json_normalize(x.to_dict(orient="records")), meta=schema)
```

but no luck. Same error.

**Comment by martindurant** [comment]
That is not surprising, as that is exactly what dask guesses as your meta. But it seems that your schema is not consistent, so you need to find one that is acceptable across all partitions, perhaps with the dtype suggestions I gave above, e.g.
```
schema["tracks.total"] = schema["tracks.total"].astype("Int64")
```
(note the upper case "I")

**Comment by frbelotto** [comment]
Is there any "ideal" solution here? Facing same issue here. A too big dataframe for pandas and need to use "json normalize"

**Comment by phofl** [comment]
The tricky part about this is that it's not clear how many columns we end up with by only looking at parts of the data (i.e. different partitions might have different columns based on the input data). We'd have to make a columns argument required, would this still help you?

**Comment by frbelotto** [comment]
Usually we might know toe expected results, but it’s no a sure. 

Can’t it run pandas nomalkze json on several small parts and concatenate them (somehow outer join on columns ) at the end ?

**Comment by frbelotto** [comment]
> Agreed with @martindurant that this would be difficult to solve in general. For simple cases, `dask.delayed` works well.
> 
> ```python
> >>> data = [{'id': 1, 'name': {'first': 'Coleen', 'last': 'Volk'}},
> ...         {'name': {'given': 'Mose', 'family': 'Regner'}},
> ...         {'id': 2, 'name': 'Faye Raker'}]
> >>> big_data = [data, data, data]
> >>> dfs = [dask.delayed(pd.io.json.json_normalize)(x) for x in big_data]
> >>> df = dd.from_delayed(dfs)
> ```
> 
> For anything more complex, moving to the dask.bag API is reasonable. Documentation writing around this would be valuable.

Hi!
Sorry to bother, but can you help me implement your example on my code?


```
import dask.dataframe as dd
import dask.delayed as dl
import pandas as pd
import numpy as np
import ast

dataframe = pd.DataFrame({
    'status' :  ['pending', 'pending','pending', 'canceled','canceled','canceled', 'confirmed', 'confirmed','confirmed'],
    'clientId' : ['A', 'B', 'C', 'A', 'D', 'C', 'A', 'B','C'],
    'partner' :  ['A', np.nan,'C', 'A',np.nan,'C', 'A', np.nan,'C'],
    'product' : ['afiliates', 'pre-paid', 'giftcard','afiliates', 'pre-paid', 'giftcard','afiliates', 'pre-paid', 'giftcard'],
    'brand' : ['brand_1', 'brand_2', 'brand_3','brand_1', 'brand_2', 'brand_3','brand_1', 'brand_3', 'brand_3'],
    'sku' : [['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio'],['rice', 'beans', 'phone', 'tv', 'alexa', 'radio']],
    'detail': [{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','extra':'take','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'},{'nickname':'todo','bestfriend':'none','eat':'meat'}],
    'gmv' : [100,100,100,100,100,100,100,100,100]
})
dataframe = dd.from_pandas(dataframe, npartitions = 3)
dataframe['sku'] = dataframe['sku'].apply(ast.literal_eval,  meta=('sku', 'object'))
dataframe['detail'] = dataframe['detail'].apply(ast.literal_eval,  meta=('sku', 'object'))
dfs = [dl.delayed(pd.json_normalize)(x) for x in dataframe['detail']] 
```

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[23], [line 6](vscode-notebook-cell:?execution_count=23&line=6)
      [1](vscode-notebook-cell:?execution_count=23&line=1) # dataframe.head()
      [2](vscode-notebook-cell:?execution_count=23&line=2) # dataframe = dataframe.explode(column='sku')
      [3](vscode-notebook-cell:?execution_count=23&line=3) # # dataframe
      [4](vscode-notebook-cell:?execution_count=23&line=4) # # dataframe.head()
      [5](vscode-notebook-cell:?execution_count=23&line=5) # df_normalized = pd.json_normalize(dataframe['detail'])
----> [6](vscode-notebook-cell:?execution_count=23&line=6) dfs = [dl.delayed(pd.json_normalize)(x) for x in dataframe['detail']]
      [7](vscode-notebook-cell:?execution_count=23&line=7) # dfs = dl.delayed(pd.json_normalize)(dataframe['detail'])
      [8](vscode-notebook-cell:?execution_count=23&line=8) # df_normalized = dl.from_delayed(dfs)

Cell In[23], [line 6](vscode-notebook-cell:?execution_count=23&line=6)
      [1](vscode-notebook-cell:?execution_count=23&line=1) # dataframe.head()
      [2](vscode-notebook-cell:?execution_count=23&line=2) # dataframe = dataframe.explode(column='sku')
      [3](vscode-notebook-cell:?execution_count=23&line=3) # # dataframe
      [4](vscode-notebook-cell:?execution_count=23&line=4) # # dataframe.head()
      [5](vscode-notebook-cell:?execution_count=23&line=5) # df_normalized = pd.json_normalize(dataframe['detail'])
----> [6](vscode-notebook-cell:?execution_count=23&line=6) dfs = [dl.delayed(pd.json_normalize)(x) for x in dataframe['detail']]
      [7](vscode-notebook-cell:?execution_count=23&line=7) # dfs = dl.delayed(pd.json_normalize)(dataframe['detail'])
      [8](vscode-notebook-cell:?execution_count=23&line=8) # df_normalized = dl.from_delayed(dfs)

AttributeError: 'curry' object has no attribute 'delayed'


Is this because my dataframe is already a dask dataframe (lazy object)? 
I´ve noticed that if I just apply the common pandas method over my dask dataframe, it does run, but how it seens to being computed on call, I am woried about memory consumption (my real dataframe is hughe!)

df_normalized = pd.json_normalize(dataframe['detail'])
df_normalized

**Comment by frbelotto** [comment]
Hi guys!
I´ve tried normalizing my dask datafame by looping over each partition, computing and saving it as a parquet file.

```
def normaliza_df() : 
    for partition in range(df.npartitions): #Carrega os dados expandidos por partição, normaliza e salva
        df_normalizado = normalizado.get_partition(partition).compute() 
        df_normalizado = pd.json_normalize(df_normalizado['historicoContabil'])
        print(f'Salvando a partição nº{partition} do normalizado')
        df_normalizado.to_parquet(path=f'Temp/Output/Normalized/Normalized_{partition}.parquet', compression='zstd', index=True)   
```

But sadly, some rows containing null values on the ['historicoContabil'] column. So, the output of the normalized file is a empty parquet and broking my code :(

**Comment by phofl** [comment]
Yeah this doesn't work since we don't know the indices. I would use join or setting an index that you want to align on

**Comment by frbelotto** [comment]
> Yeah this doesn't work since we don't know the indices. I would use join or setting an index that you want to align on

I could do it by setting the npartitions to 1, resetting the index and then merging/joining with the index.
But my issue is that changing de partition and resetting the index seems too memory expensive and crashes my virtual machine. 

How could I align the index? (resetting the index without changing the npartitions does not solve as each parquet file get a indivual index)

**Comment by frbelotto** [comment]
For those who face the same issue, so far the best way I´ve found was to create a column with a row number (cumsum) and use it to join the json_normlize output. This seems to be easier than just resetting the index, because resetting index on Dask works different that on pandas (it create an index for each partition). 

df = df.assign(idx=1)
df = df.set_index(df.idx.cumsum() - 1)
df = df.drop(columns = ['idx'])
normalized = dask.delayed(pd.json_normalize)(df['json_column'])
normalized = dd.from_delayed(normalized) # (Or you can ouput the normalized straight to a parquet file without the "from_normalized")
df = df.merge(normalizado, how='left', left_index = True , npartitions=df.npartitions , right_index=True)
